Refactor Panoptes authentication in all packages

eatyourgreens opened the pull request:
Refactor `usePanoptesUser` to get the current user from the stored Panoptes JWT (#4766.) Fall back to `auth.checkCurrent()`, which gets a new session and token from the Panoptes API, if we don't have a stored token.

Refactor the classifier to authenticate in `ClassifierContainer`, waiting until sign-in is complete and all user data has loaded before rendering the `Classifier` component (#4845 and #4875.) Move all the workflow validation up to `ClassifierContainer`, so that we validate the workflow ID (from the URL) when the user has loaded, then pass down null to `Classifier` if they don't have permission to view that workflow.

Test `ClassifierContainer`, checking that:
- inactive workflows only load when you have permission to view them.
- the mock API for `/subjects/queued` is only called once.

Fix a bug, in the project app, where updating the project user in the project app's store will unmount and mount the classifier (#4946.)

Fix a bug where a `project_roles` request is continually made (closes #4975).

The `Classifier` component uses MobX observable data, but isn't wrapped in `observer`, so I've also fixed that here.

_Please request review from `@zooniverse/frontend` team or an individual member of that team._ 

## Package
- app-content-pages
- app-project
- lib-classifier

## Linked Issue and/or Talk Post
- Closes #4766.
- Closes #4845.
- Closes #4875.
- Closes #4946.
- Closes #4975.
- https://github.com/zooniverse/panoptes-javascript-client/issues/53 

## How to Review
This PR reflects on the Panoptes JSON Web Token to get the logged-in user for the project app, the content pages app and the classifier, which should be faster than pinging the API for the current user session and won’t ask Panoptes OAuth to generate a new token (https://github.com/zooniverse/panoptes-javascript-client/issues/53.) If there's no stored token yet, it falls back to `auth.checkCurrent()`, which is currently used for all user checks.

You should be able to use any project, logged-in or not, and continue to use Admin Mode without any changes to the app's behaviour.

On page load, there should only be one request to `/api/me` to get the user resource. 

Subjects should only be fetched once, with your access token, after you log in (#4875.)

Apps will still lose your stored session when they unload eg. when going from PFE to FEM on the same origin, or when going from the project app to the content pages app. That can be tested in staging, on the `https://frontend.preview.zooniverse.org` origin, after this PR merges. 

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [x] Tests are passing locally and on Github
- [x] Documentation is up to date and changelog has been updated if appropriate
- [x] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [x] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [x] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [x] Can submit a classification
- [x] Can sign-in and sign-out
- [x] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## Refactoring
- [x] The PR creator has described the reason for refactoring
- [x] The refactored component(s) continue to work as expected

eatyourgreens commented:
Adding the 'bug' label to this because it fixes a small bug in the classifier. When you are already logged in, on the Classify page, `usePanoptesUser` in the classifier returns null when the `Classifier` component first loads.

eatyourgreens commented:
You can try out a production build of this branch here:
https://fe-project-branch.preview.zooniverse.org/projects/zooniverse/gravity-spy

goplayoutside3 commented:
In addition to the above comment, I clicked on https://fe-project-branch.preview.zooniverse.org/projects/zooniverse/gravity-spy, clicked on the first workflow and immediately saw 2 queue requests in the network tab

eatyourgreens commented:
Yeah, I traced that bug to a hook that’s running twice, so this PR won’t fix it. I’ll remove it. 

https://github.com/zooniverse/front-end-monorepo/issues/4875#issuecomment-1607787426

goplayoutside3 commented:
Is #4845 closed in this PR too? I logged the `user` from `usePanoptesUser` in the Classifier component, and when I navigate to Planet Hunters TESS's classify page, `user` is `null` while some of the other functions in the component are running. Should I be looking for something else?

<img width="330" alt="Screenshot 2023-06-27 at 1 59 34 PM" src="https://github.com/zooniverse/front-end-monorepo/assets/23665803/242086d6-4a28-4d3c-aefd-513f6c1975aa">

eatyourgreens commented:
Hmmm, that sounds like it shouldn’t happen. My reading of the auth client is that you should be logged in almost immediately once you have an access token. 

When I load the Gravity Spy Level 1 workflow, while logged in, the classifier loads and shows a subject almost immediately. 
https://fe-project-branch.preview.zooniverse.org/projects/zooniverse/gravity-spy/classify/workflow/1610

I’m not sure if it’s faster than staging. I’d expect it to be faster than if the classifier gets the user from `/api/me` when it loads. 
https://frontend.preview.zooniverse.org/projects/zooniverse/gravity-spy/classify/workflow/1610

eatyourgreens commented:
I thought that `auth.checkCurrent()` calls `/api/me` each time it runs, but that turns out not to be true. See https://github.com/zooniverse/panoptes-javascript-client/issues/207.

eatyourgreens commented:
I've rebased this, so that it's still ahead of the latest `production-release` and rewritten user auth in the classifier. The Panoptes user is now checked in `ClassifierContainer` and `usePanoptesUser` returns a `loading` flag while the API is being checked. The `Classifier` component isn't rendered until loading is complete. I think that should fix #4875.

eatyourgreens commented:
I can't reproduce either #4875 or #4946 on this branch, so I'm marking those as fixed. Checked with Gravity Spy: Level 1: https://fe-project-branch.preview.zooniverse.org/projects/zooniverse/gravity-spy/classify/workflow/1610

`ClassifierContainer` waits for sign-in to complete (either with an existing valid token or from the Panoptes API) before mounting the classifier, so #4845 should be fixed too.

eatyourgreens commented on packages/lib-classifier/src/components/Classifier/Classifier.js:
```diff
@@ -81,22 +53,20 @@ export default function Classifier({
       console.log('starting new subject queue', { workflowID, subjectSetID, subjectID })
       workflows.setResources([workflowSnapshot])
-      workflows.selectWorkflow(workflowID, subjectSetID, subjectID, canPreviewWorkflows)
+      workflows.selectWorkflow(workflowID, subjectSetID, subjectID)
     }
-  }, [canPreviewWorkflows, subjectID, subjectSetID, workflowID])
```
The `WorkflowStore` model needs updating, to either remove `canPreviewWorkflows` or set the default to `true`. Otherwise, previewing inactive workflows won't work.

eatyourgreens replied:
The tests that were removed from `Classifier` need to be moved up to `ClassifierContainer` too.

eatyourgreens commented:
Tests are passing now, so this ready for review.

goplayoutside3 commented on packages/app-project/stores/User/User.spec.js:
```diff
@@ -1,21 +1,80 @@
 import asyncStates from '@zooniverse/async-states'
+import * as client from '@zooniverse/panoptes-js'
 import { expect } from 'chai'
+import { when } from 'mobx'
 import { getSnapshot } from 'mobx-state-tree'
+import nock from 'nock'
 import sinon from 'sinon'
 
 import Store from '@stores/Store'
 import initStore from '@stores/initStore'
```
```suggestion
import { when } from 'mobx'
import nock from 'nock'

import Store from '@stores/Store'
```

goplayoutside3 commented on packages/lib-classifier/src/components/Classifier/Classifier.js:
```diff
@@ -1,33 +1,29 @@
-
-  const project = classifierStore.projects.active
+  const classifierStore = useStores()
+  const { workflows, projects } = classifierStore
+  const project = projects.active
```
```suggestion
  const { workflows } = classifierStore
```

goplayoutside3 commented on packages/lib-classifier/src/components/Classifier/ClassifierContainer.js:
```diff
@@ -1,3 +1,4 @@
+import asyncStates from '@zooniverse/async-states'
```
```suggestion
```

goplayoutside3 commented on packages/lib-classifier/src/components/Classifier/ClassifierContainer.js:
```diff
@@ -143,11 +170,15 @@ ClassifierContainer.propTypes = {
-  theme: PropTypes.object
+  subjectID: PropTypes.string,
+  subjectSetID: PropTypes.string,
+  theme: PropTypes.object,
```
```suggestion
```

`mode` is no longer a prop either.

goplayoutside3 commented on packages/lib-classifier/src/components/Classifier/ClassifierContainer.spec.js:
```diff
@@ -0,0 +1,715 @@
+          wrapper: withGrommet()
+        }
+      )
+      const taskTab = await screen.findByRole('tab', { name: 'TaskArea.task'})
+      const tutorialTab = screen.queryByRole('tab', { name: 'TaskArea.tutorial'})
```
```suggestion
```

These two are fetched eight times in the spec file, but not used. Is that intentional or might be left over from moving tests around?

eatyourgreens replied:
They're testing that the tabs have loaded, but I think that might be redundant.

eatyourgreens replied:
Removing them per your suggestion will fail the tests, but there might be a better way to wait for the workflow to have loaded and the first subject to be ready.

goplayoutside3 replied:
A note sounds good too! On code read I was just confused as to why they're grabbed from the screen, but then not used in any assertions

eatyourgreens replied:
Isn't that true of the `Classifier` tests too?

eatyourgreens replied:
Reading back through these tests, I don't test that `/subjects/queued` is only called once (#4875.) I'll fix that too.

eatyourgreens replied:
All fixed now. I added a second mock for `/subjects/queued`, and expect it not to be called.

goplayoutside3 commented on packages/lib-classifier/src/components/Classifier/hooks/usePanoptesUserSession.js:
```diff
@@ -0,0 +1,19 @@
+} from '@hooks'
+
+export default function usePanoptesUserSession({ authClient, projectID }) {
+  const { user, loading: userLoading } = usePanoptesUser(authClient)
```
Non-blocking: Because `userProjectPreferences()` and `useProjectRoles()` return an `isLoading` variable, would it make sense for `usePanoptesUserSession()` and `usePanoptesUser()` to also return `isLoading` rather than "loading" for consistency? Or perhaps the other way around changing the former two to `loading`.

eatyourgreens replied:
I've updated `usePanoptesUser` to use a similar API to [the `useSWR` API.](https://swr.vercel.app/docs/api)

goplayoutside3 commented on packages/lib-classifier/src/hooks/usePanoptesAuth.js:
```diff
@@ -3,8 +3,7 @@ import { useEffect, useState } from 'react'
 import { useStores } from '@hooks'
```
```suggestion
```

goplayoutside3 commented on packages/lib-classifier/src/store/SubjectViewerStore/SubjectViewerStore.js:
```diff
@@ -39,7 +39,7 @@ const SubjectViewer = types
   .views(self => ({
     get disableImageToolbar () {
-      const subject = getRoot(self).subjects.active
+      const subject = tryReference(() => getRoot(self).subjects?.active)
```
Question: Just for my understanding, what does the change here implement?

eatyourgreens replied:
`subjects.active` is a reference (it's actually the subject ID.) References should be accessed via `tryReference` to avoid errors. 

https://mobx-state-tree.js.org/API/#tryreference

eatyourgreens replied:
The mock store errored on this line, because `subjects.active` isn't always a valid reference.

goplayoutside3 reviewed (approved):
Looks good! I re-tested each of the Issues tagged here. The dev classifier runs as expected, as do the Zooniverse content pages and classifier storybook.

eatyourgreens commented on packages/lib-classifier/src/components/Classifier/ClassifierContainer.spec.js:
```diff
@@ -0,0 +1,715 @@
+      )
+      const taskTab = await screen.findByRole('tab', { name: 'TaskArea.task'})
+      const tutorialTab = screen.queryByRole('tab', { name: 'TaskArea.tutorial'})
+      const tabPanel = screen.queryByRole('tabpanel', { name: '1 Tab Contents'})
```
```suggestion
      const tabPanel = await screen.findByRole('tabpanel', { name: '1 Tab Contents'})
```

Maybe change this to wait for the first task to be ready?

eatyourgreens commented on packages/lib-classifier/src/hooks/usePanoptesUser.js:
```diff
@@ -20,5 +42,5 @@ export default function usePanoptesUser() {
   }, [authClient])
 
-  return user
+  return { user, loading }
```
```suggestion
  return { user, isLoading: loading }
```
Change this for consistency with `useSWR`.
